File: bin/Zextractor.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <algorithm>
#include <cmath>
#include <stdlib.h>

using namespace std;

int main(int argc, char** argv){
  ifstream hhrfile(argv[1]);
  vector<string> templates;  
  vector<double> scores;

  double total=0.0;
  for (string line; getline(hhrfile,line);){
    if (line.substr(0,1).compare(">") == 0){
      string target=line.substr(1);
      if (find(templates.begin(),templates.end(),target) != templates.end()){
	continue;
      }
      templates.push_back(target);
      getline(hhrfile,line);
      double score=atof(line.substr(line.find("Sum_probs=")+10).c_str());
      total+=score;
      scores.push_back(score);
    }
  }

  if (scores.size() != templates.size()){
    cout << "Array sizes are not equal" << endl;
    return 1;
  }

  double mean=total/scores.size();
  double stdev=0.0;
  for (vector<double>::iterator it=scores.begin(); it!=scores.end(); ++it){
    stdev+=pow(*it-mean,2.0)/scores.size();
  }
  stdev=pow(stdev,0.5);
  
  vector<string>::iterator tit=templates.begin();
  vector<double>::iterator sit=scores.begin();
  while (tit!=templates.end() || sit!=scores.end()){
    double z=((*sit-mean)/stdev);
    cout << *tit << " " << z << endl;
    ++tit;
    ++sit;
  }
}

File: bin/Zextractor.pl
use strict;
use warnings;

my $hhrfilename=$ARGV[0];
my @hits=fetchHits($hhrfilename);
for my $hit (@hits){
    my @pair=@{$hit};
    print "$pair[0] $pair[1]\n";
}

sub fetchHits{
    my $hhrfilename=$_[0];
    my @templates=();
    my @scores=();
    open(my $hhrfile,"<",$hhrfilename);
    while (my $line=<$hhrfile>){
        if (substr($line,0,1) eq ">"){
            chomp($line);
            (my $target=$line)=~s/>//g;
            next if (grep(/^$target$/,@templates));
            push(@templates,$target);
            my $scoreline=<$hhrfile>;
            $scoreline=~/Sum_probs=(\S+)/;
            my $score=$1;
            #print "$scoreline$score\n";
            push(@scores,$score);
        }
    }
    close($hhrfile);

    print scalar(@templates)."\n";
    my $meanval=0.0;
    for my $score (@scores){
        $meanval+=$score/scalar(@scores);
    }
    print "$meanval\n";
    my $std=0.0;
    for my $score (@scores){
        $std+=($score-$meanval)**2/scalar(@scores);
    }
    $std=$std**(0.5);

    my @pairs=();
    for my $i (0..scalar(@templates)-1){
        my @pair=($templates[$i],($scores[$i]-$meanval)/$std);
        push(@pairs,\@pair);
    }

    @pairs=sort{$b->[1]<=>$a->[1]} @pairs;
    return @pairs
}
